oxide-auth/src/code_grant/extensions/pkce.rs

```diff
@@ -1,11 +1,46 @@
 use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{Display, Formatter};
 
-use crate::primitives::grant::{GrantExtension, Value};
+use crate::{
+    primitives::grant::{GrantExtension, Value},
+    util::avoid_alloc_cow_str_to_string,
+};
 
 use base64;
 use sha2::{Digest, Sha256};
 use subtle::ConstantTimeEq;
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PkceError {
+    NoChallenge,
+    NoVerifier,
+    FailedVerification(&'static str),
```

Comment:
Rule of thumb, don't elaborate on verification failures in user facing context. They failed, everything beyond is a potentialy leak of information. Even if these are only ever filled with one string per method, that's not clear to a  reviewer at this point.

Should be `FailedVerification`, potentially you can add the method in some way but I don't really think it's necessary. Also, can you list this variant last?